// @tanstack/router-vite-plugin/dist/cjs/index.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const path = require("path");
const url = require("url");
const zod = require("zod");
const routerGenerator = require("@tanstack/router-generator");
const compilers = require("./compilers.cjs");
const constants = require("./constants.cjs");
const configSchema = routerGenerator.configSchema.extend({
  enableRouteGeneration: zod.z.boolean().optional(),
  experimental: zod.z.object({
    enableCodeSplitting: zod.z.boolean().optional()
  }).optional()
});
const CONFIG_FILE_NAME = "tsr.config.json";
const getConfig = async (inlineConfig, root) => {
  const config = await routerGenerator.getConfig(inlineConfig, root);
  return configSchema.parse({ ...inlineConfig, ...config });
};
function TanStackRouterVite(inlineConfig = {}) {
  return [
    TanStackRouterViteGenerator(inlineConfig),
    TanStackRouterViteCodeSplitter(inlineConfig)
  ];
}
function TanStackRouterViteGenerator(inlineConfig = {}) {
  let ROOT = process.cwd();
  let userConfig;
  const generate = async () => {
    try {
      await routerGenerator.generator(userConfig);
    } catch (err) {
      console.error(err);
      console.info();
    }
  };
  const handleFile = async (file) => {
    const filePath = path.normalize(file);
    if (filePath === path.join(ROOT, CONFIG_FILE_NAME)) {
      userConfig = await getConfig(inlineConfig, ROOT);
      return;
    }
    const routesDirectoryPath = path.isAbsolute(userConfig.routesDirectory) ? userConfig.routesDirectory : path.join(ROOT, userConfig.routesDirectory);
    if (filePath.startsWith(routesDirectoryPath)) {
      await generate();
    }
  };
  return {
    name: "vite-plugin-tanstack-router-generator",
    configResolved: async (config) => {
      ROOT = config.root;
      userConfig = await getConfig(inlineConfig, ROOT);
      if (userConfig.enableRouteGeneration ?? true) {
        await generate();
      }
    },
    watchChange: async (file, context) => {
      if (userConfig.enableRouteGeneration ?? true) {
        if (["create", "update", "delete"].includes(context.event)) {
          await handleFile(file);
        }
      }
    }
  };
}
function fileIsInRoutesDirectory(filePath, routesDirectory) {
  const routesDirectoryPath = path.isAbsolute(routesDirectory) ? routesDirectory : path.join(process.cwd(), routesDirectory);
  return filePath.startsWith(routesDirectoryPath);
}
function TanStackRouterViteCodeSplitter(inlineConfig = {}) {
  let ROOT = process.cwd();
  let userConfig;
  return {
    name: "vite-plugin-tanstack-router-code-splitter",
    enforce: "pre",
    configResolved: async (config) => {
      ROOT = config.root;
      userConfig = await getConfig(inlineConfig, ROOT);
    },
    resolveId(source) {
      var _a;
      if (!((_a = userConfig.experimental) == null ? void 0 : _a.enableCodeSplitting)) {
        return null;
      }
      if (source.startsWith(constants.splitPrefix + ":")) {
        return source.replace(constants.splitPrefix + ":", "");
      }
      return null;
    },
    async transform(code, id, transformOptions) {
      var _a;
      if (!((_a = userConfig.experimental) == null ? void 0 : _a.enableCodeSplitting)) {
        return null;
      }
      const url$1 = url.pathToFileURL(id);
      url$1.searchParams.delete("v");
      id = url.fileURLToPath(url$1).replace(/\\/g, "/");
      const compile = compilers.makeCompile({
        root: ROOT
      });
      if (id.includes(constants.splitPrefix)) {
        const compiled = await compilers.splitFile({
          code,
          compile,
          filename: id
          // ref,
        });
        return compiled;
      } else if (fileIsInRoutesDirectory(id, userConfig.routesDirectory) && (code.includes("createRoute(") || code.includes("createFileRoute("))) {
        if (code.includes("@react-refresh")) {
          throw new Error(
            `We detected that the '@vitejs/plugin-react' was passed before '@tanstack/router-vite-plugin'. Please make sure that '@tanstack/router-vite-plugin' is passed before '@vitejs/plugin-react' and try again: 
e.g.

plugins: [
  TanStackRouterVite(), // Place this before viteReact()
  viteReact(),
]
`
          );
        }
        const compiled = await compilers.compileFile({
          code,
          compile,
          filename: id
        });
        return compiled;
      }
      return null;
    }
  };
}
exports.TanStackRouterVite = TanStackRouterVite;
exports.TanStackRouterViteCodeSplitter = TanStackRouterViteCodeSplitter;
exports.TanStackRouterViteGenerator = TanStackRouterViteGenerator;
exports.configSchema = configSchema;
